netty/src/test/java/io/grpc/netty/ProtocolNegotiatorsTest.java

```diff
@@ -160,6 +167,10 @@ public static void loadCerts() throws Exception {
   private SSLEngine engine;
   private ChannelHandlerContext channelHandlerCtx;
 
+  private InternalLogger mockLogger4KeyLog;
+  private InternalLogger mockLogger4Others;
```

Comment:
Instead of returning this `mockLogger4Others`, how about the factory delegate to `oldLoggerFactory`?

Honestly, even for mockLogger4KeyLog it'd probably be better to use `AdditionalAnswers.delegatesTo()` and delegate to a real instance. That gets you the recording capability of the mock while not needing to define Answers. Right now if the code under test did things like calling `isWarnEnabled()` it would be returned `false` and so may not log at all. That would fail one of the tests, but it is still good to back mocks with real objects for anything other than Listeners.